### ISSUE TITLE
Error backpropagation in Keras

### ISSUE THREAD

**Issue by tkrishp** [fchollet/keras#2566]
Hi,

I'm new to Keras and currently evaluating the library for my research purpose. Could someone clarify if Keras supports error backpropagation to train a neural network built using DenseLayers? I could not find any references in the documentation. I'm working on a multi-class classification problem and the objective function is 'categorical_crossentropy'.

**Comment by joelthchao** [comment]
Please reference to [keras/examples](https://github.com/fchollet/keras/blob/master/examples) folder, there are lots of good examples.
For example, [mnist example](https://github.com/fchollet/keras/blob/master/examples/mnist_mlp.py) shows you how to build a multi-class classification problem.

**Comment by tkrishp** [comment]
It's not clear from the example if the model builds a feedforward NN or uses error backpropagation. That's what I would like to get clarified.

**Comment by jocicmarko** [comment]
I think that your lack of knowledge on the subject makes you misunderstand the difference between terms "feed-forward neural networks" and "error backpropagation".

Feed-forward NNs are just a type of NNs that propagate the signal in one direction (forward), so to speak - examples are multi-layer perceptrons, convolutional networks... On the other hand, error backpropagation is a general term for an approach for training neural networks.

**Comment by BaldwinTheThird** [comment]
I think he is aware of the difference between FFNN and BP, when he says "or" he actually means "and whether". To answer the question. My understanding is that Keras is built on top of Theano which employs automatic differentiation for whatever model you set up (including FFNN which would be anything where you stack non recurrent layers on top of each other). BP is nothing more than an application of the chain rule which enables you build up derivatives w.r.t. all weights in the network recursively which saves time compared to naively computing them using other techniques. Whatever Theano does ultimately translates into something akin to BP. But why worry about that? The whole point of Theano is that you no longer have to think about this part of your problem, just treat it as a given working technology, just like you treat BLAS.

**Comment by tkrishp** [comment]
@BaldwinTheThird that helps, thanks.

**Comment by jemshit** [comment]
So, how do we make sure Dense layer uses Backpropagation and changes weights according to that? I didn't see any Backpropagation+FFNN example.

**Comment by joelthchao** [comment]
Fundamental idea is that Keras use Theano/Tensorflow's auto gradient computation. If you want to understand how Keras use BP to compute gradients and update weights (no limit to any kind of model including FFNN), please refer to its [code](https://github.com/fchollet/keras/blob/master/keras/engine/training.py#L997).
